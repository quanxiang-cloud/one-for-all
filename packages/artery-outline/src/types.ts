import { HTMLNode, ReactComponentNode } from '@one-for-all/artery';

export type NodePrimary =
  | Pick<HTMLNode, 'type' | 'name'>
  | Pick<ReactComponentNode, 'type' | 'packageName' | 'packageVersion' | 'exportName'>;

export type CollapsedStatus = 'none' | 'collapsed' | 'expended';

export interface Entry {
  id: string;
  name: string;
  depth: number;
  isContainer: boolean;
}
