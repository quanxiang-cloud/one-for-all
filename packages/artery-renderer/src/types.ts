import React from 'react';
import type { BehaviorSubject } from 'rxjs';
import type { FetchParams, APISpecAdapter, AjaxConfig } from '@one-for-all/api-spec-adapter';
import type { BrowserHistory, Location } from 'history';
import type { APILoadingProperty, ConstantProperty } from '@one-for-all/artery';
import ArterySpec from '@one-for-all/artery';

export type VersatileFunc<T = unknown> = (...args: unknown[]) => T;

/**
 * Convertor is used to transform the state, such as API Result, before passing it to node.
 *
 * Convertor will NOT be called if the state is nullish, for example API request failed or the result is `null`.
 */
export type StateConvertor = (v: unknown) => unknown;

/**
 * LifecycleHooks provide a opportunity to executive some side-effects when node
 * `didMount` and `willUnmount`.
 */
export type LifecycleHooks = Partial<{
  didMount: VersatileFunc;
  willUnmount: VersatileFunc;
}>;

/**
 * toProps function should return Record<string, unknown>;
 */
export type ToProps = (state: unknown) => Record<string, unknown>;

export type NodeProperty =
  | ConstantProperty
  | APIResultProperty
  | APILoadingProperty
  | SharedStateProperty
  | NodeStateProperty
  | FunctionalProperty
  | SharedStateMutationProperty
  | APIInvokeProperty
  | RenderProperty
  | ComputedProperty
  | InheritedProperty;

export type PlainState = APIResultProperty | SharedStateProperty | NodeStateProperty | ConstantProperty;

/**
 * APIResultProperty define a value converted from API response.
 */
export interface APIResultProperty extends Omit<ArterySpec.APIResultProperty, 'convertor'> {
  convertor?: StateConvertor;
}

export interface SharedStateProperty extends Omit<ArterySpec.SharedStateProperty, 'convertor'> {
  convertor?: StateConvertor;
}

export interface NodeStateProperty extends Omit<ArterySpec.NodeStateProperty, 'convertor'> {
  convertor?: StateConvertor;
}

export interface FunctionalProperty extends Omit<ArterySpec.FunctionalProperty, 'func'> {
  func: VersatileFunc;
}

/**
 * @deprecated This type has been deprecated, please use FunctionalProperty instead
 */
export interface SharedStateMutationProperty
  extends Omit<ArterySpec.SharedStateMutationProperty, 'convertor'> {
  convertor?: VersatileFunc;
}

/**
 * @deprecated This type has been deprecated, please use FunctionalProperty instead
 */
export interface APIInvokeProperty extends Omit<ArterySpec.APIInvokeProperty, 'paramsBuilder' | 'callback'> {
  // the required return type is too complex
  paramsBuilder?: (...args: unknown[]) => FetchParams;
  callback?: APIFetchCallback;
}

// It is common to passing a component to another component,
// or defined a function which may take some argument and return a component.
// RenderProperty is the spec for this kind of property.
//
// <ParentComponent
//   render={<SomeComponent />}
// />
// <ParentComponent
//   render={(someData): JSX.Element => (<SomeComponent data={someData} />)}
// />
// <ParentComponent
//   render={(someData, someIgnoredValue): JSX.Element => (<SomeComponent data={someData} otherProp={otherProp} />)}
// />
export interface RenderProperty extends Omit<ArterySpec.RenderProperty, 'adapter' | 'node'> {
  adapter: (...args: unknown[]) => Record<string, unknown>;
  node: ArteryNode;
}

export interface ComputedProperty extends Omit<ArterySpec.ComputedProperty, 'convertor'> {
  convertor: StateConvertor;
}

export interface InheritedProperty extends Omit<ArterySpec.InheritedProperty, 'convertor'> {
  convertor?: StateConvertor;
}

export type NodeProperties = Record<string, NodeProperty>;

export type ArteryNode =
  | HTMLNode
  | LinkNode
  | ReactComponentNode
  | LoopContainerNode
  | ComposedNode
  | RefNode
  | JSXNode
  | RouteNode;

export type ShouldRenderCondition =
  | APIResultProperty
  | NodeStateProperty
  | SharedStateProperty
  | ComputedProperty
  | (ArterySpec.APILoadingProperty & { revert?: boolean });

export interface BaseNode extends Omit<ArterySpec.BaseNode, 'props' | 'shouldRender' | 'lifecycleHooks'> {
  props?: NodeProperties;
  shouldRender?: ShouldRenderCondition;
  lifecycleHooks?: LifecycleHooks;
}

export interface HTMLNode extends BaseNode, Pick<ArterySpec.HTMLNode, 'name'> {
  type: 'html-element';
  children?: Array<ArteryNode>;
}

export interface LinkNode extends HTMLNode {
  name: 'a';
  isLink: true;
}

export interface ReactComponentNode
  extends BaseNode,
    Pick<
      ArterySpec.ReactComponentNode,
      'packageName' | 'packageVersion' | 'exportName' | 'supportStateExposure'
    > {
  type: 'react-component';
  children?: Array<ArteryNode>;
}

export interface RouteNode extends BaseNode {
  type: 'route-node';
  path: string;
  node: ArteryNode;
  exactly?: boolean;
}

export interface IndividualLoopContainer extends BaseNode, Pick<ArterySpec.LoopContainerNode, 'loopKey'> {
  type: 'loop-container';
  iterableState: PlainState;
  node: ArteryNode;
  toProps: ToProps;
}

export interface ComposedNodeLoopContainer extends BaseNode, Pick<ArterySpec.LoopContainerNode, 'loopKey'> {
  type: 'loop-container';
  iterableState: PlainState;
  node: ComposedNode;
}

export type LoopContainerNode = IndividualLoopContainer | ComposedNodeLoopContainer;

export type ComposedNodeChild = ArteryNode & {
  toProps: ToProps;
};

export type ComposeOutLayer = Omit<HTMLNode, 'children'> | Omit<ReactComponentNode, 'children'>;

export interface ComposedNode extends BaseNode {
  type: 'composed-node';
  outLayer?: ComposeOutLayer;
  children?: Array<ComposedNodeChild>;
  nodes: Array<ComposedNodeChild>;
}

export interface RefNode extends BaseNode, Pick<ArterySpec.RefNode, 'arteryID' | 'orphan'> {
  type: 'ref-node';
  fallback?: ArteryNode;
}

export interface JSXNode extends BaseNode, Pick<ArterySpec.JSXNode, 'jsx'> {
  type: 'jsx-node';
}

export interface FetchOption {
  params?: FetchParams;
  // Callback is the hook for performing side effect after an API request.
  //
  // Why use callback style hook instead of Promise style?
  //
  // Every node should be equal, but some nodes are more equal than others.
  // In Pro-Code scenario, we can carefully design a Model to handle view actions and perform API request,
  // we also can carefully design some FLAG state to prevent extra API request or an complex cancellation mechanism.
  // These solutions may not be difficult for experienced programmers,
  // but our goal of building low-code platform is to enable junior programmers, even non-programmers,
  // to develop complex applications quickly and easily.
  // To solve this problem, we use `rxjs/ajax` to implement automatic cancellation of requests.
  // Doing so led us to a small problem, we can not use Promise style hook.
  // As said at the beginning, every node should be equal,
  // so every node could perform a side effect after some API request,
  // and if we use a Promise style hook,
  // there must will be some hooks will never be called and stay in pending status forever,
  // just like the following code:
  //
  //   new Promise((resolve, reject) => {
  //     setTimeout(() => {
  //       // resolve will never be called
  //       resolve();
  //     }, forever);
  //   });
  //
  // We, the developers, design a action which has a side effect,
  // but user make the finial decision when to trigger that action.
  // Side effect could be perform as many times as wanted,
  // but only under user's expectation.
  // When user trigger the *same action* many times,
  // only the last one should perform side effect.
  //
  // ps. Same actions means they have the same API stateID.
  callback?: APIFetchCallback;
}

export type RawFetchOption = Pick<AjaxConfig, 'method' | 'url' | 'body' | 'headers'>;

// APIState define the type of API results from view perspective.
// This type is inspired by [react-query](https://react-query.tanstack.com/).
export interface APIState {
  loading: boolean;
  result?: unknown;
  error?: Error;
}

export interface StatesHubAPI {
  hasState$: (stateID: string) => boolean;
  findState$: (stateID: string) => APIState$WithActions | undefined;
  getState$: (stateID: string) => BehaviorSubject<APIState>;
  fetch: (stateID: string, fetchOption: FetchOption) => void;
  rawFetch: (stateID: string, rawFetchOption: RawFetchOption & { callback?: APIFetchCallback }) => void;
  refresh: (stateID: string) => void;
}

export interface APIState$WithActions {
  state$: BehaviorSubject<APIState>;
  fetch: (fetchOption: FetchOption) => void;
  refresh: () => void;
  rawFetch: (rawFetchOption: RawFetchOption & { callback?: APIFetchCallback }) => void;
}

export type APIFetchCallback = (state: Omit<APIState, 'loading'>) => void;
export type APIFetch = (params: FetchParams, callback?: APIFetchCallback) => void;

export interface StatesHubShared {
  hasState$: (stateID: string) => boolean;
  findState$: (stateID: string) => BehaviorSubject<unknown> | undefined;
  getState$: (stateID: string) => BehaviorSubject<unknown>;
  getNodeState$: (nodePath: string) => BehaviorSubject<unknown>;
  exposeNodeState: (nodePath: string, state: unknown) => void;
  mutateState: (stateID: string, state: unknown) => void;
}

export interface APIStateWithFetch extends APIState {
  fetch: APIFetch;
  rawFetch: (rawFetchOption: RawFetchOption, callback?: APIFetchCallback) => void;
  refresh: () => void;
}

export interface NodePropsCache {
  getProps$: (parentID: string) => BehaviorSubject<Record<string, unknown>> | undefined;
  setProps: (path: string, nodeID: ArteryNode['id'], props: Record<string, unknown>) => void;
  shouldCache: (nodeID: string) => boolean;
}
export interface CTX {
  statesHubAPI: StatesHubAPI;
  statesHubShared: StatesHubShared;
  apiStates: Readonly<Record<string, APIStateWithFetch>>;
  states: Record<string, unknown>;
  location$: BehaviorSubject<Location>;
  history: BrowserHistory;
  nodePropsCache: NodePropsCache;

  plugins: Plugins;
}

export type ArteryRendererCTX = Pick<CTX, 'states' | 'apiStates'> & {
  history: BrowserHistory;
};

// map of stateID and apiID
export type APIStatesSpec = Record<string, { apiID: string; [key: string]: unknown }>;

export type InitializerFunc = (dependencies: Record<string, unknown>) => Promise<unknown> | unknown;

interface Initializer {
  func: InitializerFunc;
  dependencies?: {
    [key: string]: FetchParams;
  };
}

export type SharedState = Omit<ArterySpec.SharedState, 'initializer'> & {
  initializer?: Initializer;
};

export type SharedStatesSpec = Record<string, SharedState>;

export interface Artery {
  node: ArteryNode;
  apiStateSpec?: APIStatesSpec;
  sharedStatesSpec?: SharedStatesSpec;
}

// todo fixme
export type DynamicComponent = React.FC<any> | React.ComponentClass<unknown>;

type PackageNameVersion = string;
export type Repository = Record<PackageNameVersion, Record<string, DynamicComponent>>;

export type RefLoader = (arteryID: string) => Promise<{ artery: ArterySpec.Artery; plugins?: Plugins }>;

export interface ComponentLoaderParam {
  packageName: string;
  packageVersion: string;
  exportName: string;
}

export type ComponentLoader = (locator: ComponentLoaderParam) => Promise<DynamicComponent>;

export interface Plugins {
  apiSpecAdapter?: APISpecAdapter;
  repository?: Repository;
  refLoader?: RefLoader;
  componentLoader?: ComponentLoader;
}
