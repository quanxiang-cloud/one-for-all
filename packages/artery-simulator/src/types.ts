import type { Rect, ElementRect } from '@one-for-all/elements-radar';

export interface VisibleNode extends ElementRect {
  id: string;
  absolutePosition: Rect;
}

export interface SimulatorReport {
  visibleNodes: VisibleNode[];
  areaHeight: number;
  areaWidth: number;
}

export interface ShadowNode extends VisibleNode {
  nodePath: string[];
  depth: number;
  supportChildren: boolean;
  area: number;
}

export type Position =
  | 'top'
  | 'right'
  | 'bottom'
  | 'left'
  | 'inner'
  | 'inner-top'
  | 'inner-right'
  | 'inner-bottom'
  | 'inner-left';

export interface GreenZone {
  hoveringNodeID: string;
  position: Position;
  mostInnerNode: ShadowNode;
}
