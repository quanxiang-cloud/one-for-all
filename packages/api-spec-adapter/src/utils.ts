import type { Spec, Operation } from './swagger-schema-official';

import { METHODS } from './types';

export function indexOperation(spec: Spec): Record<string, Operation> {
  const operationMap: Record<string, Operation> = {};
  for (const [pathName, pathObj] of Object.entries(spec.paths || {})) {
    if (!pathObj) {
      continue;
    }

    for (const method of METHODS) {
      const operationObject = pathObj[method];
      if (!operationObject) {
        continue;
      }

      operationMap[`${method}:${pathName}`] = operationObject;
    }
  }

  return operationMap;
}

// copied from https://gist.github.com/creationix/7435851#gistcomment-3698888
// return an absolute path
export function join(...segments: string[]): string {
  const parts = segments.reduce<string[]>((parts, segment) => {
    // Remove leading slashes from non-first part.
    // Remove trailing slashes.
    const fragment = segment.replace(/^\//, '').replace(/\/$/, '');
    if (!fragment) {
      return parts;
    }

    const fragments = fragment.split('/');
    return parts.concat(fragments);
  }, []);

  const resultParts = [];
  for (const part of parts) {
    if (part === '.') {
      continue;
    }
    if (part === '..') {
      resultParts.pop();
      continue;
    }
    resultParts.push(part);
  }
  return `/${resultParts.join('/')}`;
}

type KeyPathPair = [string, string | number | boolean | null];

function isPrimaryValue(v: unknown): boolean {
  return typeof v === 'boolean' ||
    typeof v === 'number' ||
    typeof v === 'string' ||
    v === null;
}

function shouldGoIn(v: unknown): boolean {
  return typeof v === 'object' && v !== null;
}

function _toKeyPathPair(v: unknown, pairs: Array<KeyPathPair>, parentPath: string): Array<KeyPathPair> {
  if (typeof v !== 'object' || v === null) {
    return [];
  }

  Object.entries(v).forEach(([key, value]) => {
    const keyPath = parentPath ? `${parentPath}.${key}` : key;
    if (isPrimaryValue(value)) {
      pairs.push([keyPath, value]);
      return;
    }

    if (!shouldGoIn(v)) {
      return;
    }

    _toKeyPathPair((v as any)[key], pairs, keyPath);
  });

  return pairs;
}

export function toKeyPathPair(v: unknown, prefix: string): Array<KeyPathPair> {
  return _toKeyPathPair(v, [], prefix);
}
