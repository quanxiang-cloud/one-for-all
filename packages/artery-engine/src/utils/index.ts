export * from './build'
export * from './predicate';
