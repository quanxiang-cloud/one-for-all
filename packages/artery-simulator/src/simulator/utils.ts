import { parentIdsSeq } from '@one-for-all/artery-utils';
import { useEffect, useState } from 'react';
import { BehaviorSubject } from 'rxjs';
import { removeIn } from 'immutable';
import { generateNodeId } from '@one-for-all/artery-engine';
import type { Node } from '@one-for-all/artery';
import {
  byArbitrary,
  _appendTo,
  ImmutableNode,
  _insertLeftSiblingTo,
  _insertRightSiblingTo,
  ComposedNode,
  travel,
} from '@one-for-all/artery-utils';

import type { GreenZone, Cursor, Position, NodePrimary } from '../types';

const img = new Image();
img.src = 'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';

export function overrideDragImage(dateTransfer: DataTransfer): void {
  dateTransfer.setDragImage(img, 0, 0);
}

interface GetPositionParam {
  cursor: Cursor;
  hoveringRect: DOMRectReadOnly;
  supportInner: boolean;
}

// TODO optimize this
export function calcHoverPosition({ cursor, hoveringRect, supportInner }: GetPositionParam): Position {
  const leftDistance = Math.abs(cursor.x - hoveringRect.left);
  const rightDistance = Math.abs(cursor.x - hoveringRect.right);
  if (!supportInner) {
    return leftDistance < rightDistance ? 'left' : 'right';
  }

  if (leftDistance <= 9) {
    return 'left';
  }

  if (rightDistance <= 9) {
    return 'right';
  }

  const oneThirdWidth = hoveringRect.width / 3;
  if (leftDistance < oneThirdWidth) {
    return 'inner-left';
  }

  if (rightDistance < oneThirdWidth) {
    return 'inner-right';
  }

  return 'inner';
}

export function isChildNodeOf(
  root: Immutable.Collection<unknown, unknown>,
  parentID: string,
  childID: string,
): boolean {
  const parentIDs = parentIdsSeq(root, childID);
  if (!parentIDs) {
    return false;
  }

  return parentIDs.keyOf(parentID) !== undefined ? true : false;
}

export function getNodeExecutor(node: NodePrimary): string {
  if (node.type === 'html-element') {
    return `html-element:${node.name}`;
  }

  if (node.type === 'react-component') {
    return `react_component:${node.packageName}:${node.packageVersion}:${node.exportName}`;
  }

  return '';
}

export function useBehaviorSubjectState<T>(subject: BehaviorSubject<T>): T {
  const [state, setState] = useState(subject.value);

  useEffect(() => {
    const subscription = subject.subscribe(setState);
    return () => {
      subscription.unsubscribe();
    };
  }, [subject]);

  return state;
}

export function jsonParse<T>(json: string): T | undefined {
  try {
    return JSON.parse(json);
  } catch (error) {
    return;
  }
}

interface MoveNodeParams {
  rootNode: ImmutableNode;
  nodeID: string;
  greenZone: GreenZone;
}

export function moveNode({ rootNode, nodeID, greenZone }: MoveNodeParams): ImmutableNode | undefined {
  let _rootNode: ImmutableNode = rootNode;
  const nodeToMoveKeyPath = byArbitrary(rootNode, nodeID);
  if (!nodeToMoveKeyPath) {
    return;
  }

  const nodeToMove = rootNode.getIn(nodeToMoveKeyPath) as ImmutableNode | undefined;
  if (!nodeToMove) {
    return;
  }

  _rootNode = removeIn(rootNode, nodeToMoveKeyPath);

  return insertNode({ rootNode: _rootNode, node: nodeToMove, greenZone });
}

interface InsertNodeParams {
  rootNode: ImmutableNode;
  node: Node | ImmutableNode;
  greenZone: GreenZone;
}

export function insertNode({ rootNode, node, greenZone }: InsertNodeParams): ImmutableNode | undefined {
  if (
    greenZone.type === 'node_without_children' &&
    (greenZone.position === 'inner' ||
      greenZone.position === 'inner-left' ||
      greenZone.position === 'inner-right')
  ) {
    return _appendTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'node_without_children' && greenZone.position === 'left') {
    return _insertLeftSiblingTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'node_without_children' && greenZone.position === 'right') {
    return _insertRightSiblingTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'adjacent-with-parent' && greenZone.edge === 'left') {
    return _insertLeftSiblingTo(rootNode, greenZone.child.id, node);
  }

  if (greenZone.type === 'adjacent-with-parent' && greenZone.edge === 'right') {
    return _insertRightSiblingTo(rootNode, greenZone.child.id, node);
  }

  if (greenZone.type !== 'between-nodes') {
    return;
  }

  if (greenZone.left.absolutePosition.height < greenZone.right.absolutePosition.height) {
    return _insertRightSiblingTo(rootNode, greenZone.left.id, node);
  }

  if (greenZone.left.absolutePosition.height > greenZone.right.absolutePosition.height) {
    return _insertLeftSiblingTo(rootNode, greenZone.right.id, node);
  }

  return _insertRightSiblingTo(rootNode, greenZone.left.id, node);
}

function regenerateNodeID<T extends Node | ComposedNode>(node: T): T {
  node.id = generateNodeId(node.type);

  return node;
}

// todo optimize performance
export function duplicateNode(node: Node): Node {
  const newNode = travel(node, {
    htmlNode: (current) => regenerateNodeID(current),
    reactComponentNode: (current) => regenerateNodeID(current),
    loopContainerNode: (current) => regenerateNodeID(current),
    composedNode: (current) => regenerateNodeID(current),
    refNode: (current) => regenerateNodeID(current),
    jsxNode: (current) => regenerateNodeID(current),
    routeNode: (current) => regenerateNodeID(current),
  });

  return newNode;
}
