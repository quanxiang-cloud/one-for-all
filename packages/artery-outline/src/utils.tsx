import { useMemo } from 'react';
import { Node } from '@one-for-all/artery';
import {
  ancestors,
  ImmutableNode,
  KeyPath,
  _flat,
  _insertChildAt,
  _insertRightSiblingTo,
} from '@one-for-all/artery-utils';
import { fromJS, List } from 'immutable';
import { arrayMove } from '@dnd-kit/sortable';
import { logger } from '@one-for-all/utils';

import { INDENTATION_WIDTH } from './constants';
import { Entry, NodePrimary } from './types';

const NODE_TYPE_WHITE_LIST = ['html-element', 'react-component', 'loop-container'];

export function useFlattenNodes(
  rootNode: Node,
  collapsedNodeIDs: string[],
  draggingId?: string,
): Array<[KeyPath, ImmutableNode]> {
  const rawPairs = useMemo(() => {
    let pairs: Array<[KeyPath, ImmutableNode]> = _flat(fromJS(rootNode)).slice();
    for (let i = 0; i < pairs.length; ) {
      const nodeType = pairs[i][1].getIn(['type']);
      if (NODE_TYPE_WHITE_LIST.includes(nodeType as string)) {
        i = i + 1;
        continue;
      }

      const keyPathStr: string = pairs[i][0].toJS().join('/');
      while (pairs[i]?.[0].toJS().join('/').startsWith(keyPathStr)) {
        pairs.splice(i, 1);
      }
    }

    return pairs;
  }, [rootNode]);

  return useMemo(() => {
    let _rawPairs = rawPairs.slice();
    const idsShouldRemoveChildren = draggingId
      ? [draggingId, ...Array.from(collapsedNodeIDs)]
      : Array.from(collapsedNodeIDs);

    for (const id of idsShouldRemoveChildren) {
      const parentIndex = _rawPairs.findIndex(([_, n]) => n.getIn(['id']) === id);
      if (parentIndex === -1) {
        continue;
      }

      const [keyPath] = _rawPairs[parentIndex];
      const keyPathStr = keyPath.toJS().join('/');
      let lastDeleteIndex = parentIndex;
      for (let i = lastDeleteIndex + 1; i < _rawPairs.length; i += 1) {
        if (_rawPairs[i][0].toJS().join('/').startsWith(keyPathStr)) {
          lastDeleteIndex = i;
          continue;
        }

        break;
      }

      _rawPairs.splice(parentIndex + 1, lastDeleteIndex - parentIndex);
    }

    return _rawPairs;
  }, [draggingId, collapsedNodeIDs, rawPairs]);
}

export function pairToEntry(
  [keyPath, n]: [KeyPath, ImmutableNode],
  isContainer: (node: NodePrimary) => boolean,
): Entry {
  const id = n.getIn(['id']) as string;
  const primaryNode: NodePrimary = {
    type: n.getIn(['type']),
    name: n.getIn(['name']),
    packageName: n.getIn(['packageName']),
    packageVersion: n.getIn(['packageVersion']),
    exportName: n.getIn(['exportName']),
  } as NodePrimary;

  const _keyPath = keyPath.toArray().filter((path) => path === 'children' || path === 'nodes');

  return {
    id: id,
    name: (n.getIn(['label']) as string) || id,
    depth: _keyPath.length,
    isContainer: isContainer(primaryNode),
  };
}

export function getProjectionDepth(
  entries: Entry[],
  draggingId: string,
  overId: string,
  dragOffset: number,
): number {
  const overEntryIndex = entries.findIndex(({ id }) => id === overId);
  const draggingEntryIndex = entries.findIndex(({ id }) => id === draggingId);
  const draggingEntry = entries[draggingEntryIndex];
  const newEntries = arrayMove(entries, draggingEntryIndex, overEntryIndex);

  const previousEntry = newEntries[overEntryIndex - 1];
  const nextEntry = newEntries[overEntryIndex + 1];
  const dragDepth = Math.round(dragOffset / INDENTATION_WIDTH);
  const projectedDepth = draggingEntry.depth + dragDepth;

  const maxDepth = getMaxDepth(previousEntry);
  const minDepth = nextEntry ? nextEntry.depth : 0;

  let depth = projectedDepth;
  if (projectedDepth >= maxDepth) {
    depth = maxDepth;
  } else if (projectedDepth < minDepth) {
    depth = minDepth;
  }

  return depth;
}

function getMaxDepth(entry?: Entry): number {
  if (!entry) {
    return 0;
  }

  return entry.isContainer ? entry.depth + 1 : entry.depth;
}

export function insertBelowTo(
  rootNode: ImmutableNode,
  targetEntry: Entry,
  projectedDepth: number,
  node: ImmutableNode,
): ImmutableNode | undefined {
  if (projectedDepth === targetEntry.depth) {
    logger.debug(`[artery-outline] move [${node.getIn(['id'])}] to right of [${targetEntry.id}]`);

    return _insertRightSiblingTo(rootNode, targetEntry.id, node);
  }

  if (projectedDepth > targetEntry.depth) {
    logger.debug(
      `[artery-outline] move [${node.getIn(['id'])}] to the first position of [${targetEntry.id}]'s children`,
    );

    return _insertChildAt(rootNode, targetEntry.id, 0, node);
  }

  const parentIndex = targetEntry.depth - projectedDepth - 1;
  const parentList = ancestors(rootNode, targetEntry.id) || List();
  if (!parentList || !parentList.size) {
    logger.error(`[artery-outline] fatal error, can not find over node parents`);
    return;
  }

  const parentKeyPath = parentList.get(parentIndex);
  if (!parentKeyPath) {
    logger.error(`[artery-outline] fatal error, no ancestor at index ${parentIndex}`);
    return;
  }

  logger.debug(`[artery-outline] move [${node.getIn(['id'])}] to right of [${parentKeyPath.toJS()}]`);

  return _insertRightSiblingTo(rootNode, parentKeyPath, node);
}
