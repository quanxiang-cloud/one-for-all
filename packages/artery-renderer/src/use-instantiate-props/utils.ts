import { BehaviorSubject, combineLatestWith, map, of, skip, tap } from 'rxjs';
import { logger } from '@one-for-all/utils';
import type { ComputedDependency } from '@one-for-all/artery';

import { CTX, StateConvertor } from '../types';

interface ConvertResultParams {
  state: unknown;
  convertor?: StateConvertor;
  fallback: unknown;
  propName: string;
}

export function convertState({ state, convertor, fallback, propName }: ConvertResultParams): unknown {
  if (convertor && state !== undefined) {
    try {
      return convertor(state) ?? fallback;
    } catch (error) {
      logger.error(
        `an error occurred while calling state convertor for prop: "${propName}"`,
        'with the following state and convertor:',
        '\n',
        state,
        '\n',
        convertor.toString(),
        '\n',
        'So return fallback instead, fallback:',
        fallback,
        '\n',
        '\n',
        error,
      );
      return fallback;
    }
  }

  return state ?? fallback;
}

interface GetComputedState$Props {
  propName: string;
  deps: ComputedDependency[];
  convertor: StateConvertor;
  ctx: CTX;
  fallback: unknown;
}

export function getComputedState$({
  propName,
  deps,
  convertor,
  ctx,
  fallback,
}: GetComputedState$Props): BehaviorSubject<unknown> {
  let _fallback = fallback;
  const deps$ = deps.map<BehaviorSubject<unknown>>(({ type, depID }) => {
    if (type === 'api_state') {
      return ctx.statesHubAPI.getState$(depID) as BehaviorSubject<unknown>;
    }

    if (type === 'node_state') {
      return ctx.statesHubShared.getNodeState$(depID);
    }

    return ctx.statesHubShared.getState$(depID);
  });
  const initialDeps = deps$.reduce((acc: Record<string, unknown>, dep$, index) => {
    const key = deps[index].depID;
    acc[key] = dep$.value;
    
    return acc;
  }, {});
  const state$ = new BehaviorSubject(
    convertState({
      state: initialDeps,
      convertor: convertor,
      fallback: fallback,
      propName,
    }),
  );

  of(true)
    .pipe(
      combineLatestWith(deps$),
      map((_, ..._dep) => {
        return convertState({
          state: _dep,
          convertor: convertor,
          fallback: _fallback,
          propName,
        });
      }),
      skip(1),
      tap((state) => {
        _fallback = state;
      }),
    )
    .subscribe((state) => state$.next(state));

  return state$;
}
