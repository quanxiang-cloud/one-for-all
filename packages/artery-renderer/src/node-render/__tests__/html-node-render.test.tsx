import React from 'react';
import { logger } from '@one-for-all/utils';
import { render } from '@testing-library/react';

import HTMLNodeRender from '../html-node-render';
import dummyCTX from '../../boot-up/__tests__/fixtures/dummy-ctx';
import { HTMLNode } from '../../types';

test('HTMLNodeRender_should_handle_empty_html_tag', () => {
  const node: HTMLNode = {
    id: 'some_node_id',
    type: 'html-element',
    name: '',
    props: {
      id: {
        type: 'constant_property',
        value: 'some_id',
      },
      className: {
        type: 'constant_property',
        value: 'foo bar',
      },
    },
  };

  const { container } = render(<HTMLNodeRender node={node} ctx={dummyCTX} />);

  expect(logger.error).toBeCalled();
  expect(container).toMatchSnapshot();
});

test('HTMLNodeRender_match_snapshots', () => {
  const node: HTMLNode = {
    id: 'some_node_id',
    type: 'html-element',
    name: 'div',
    props: {
      id: {
        type: 'constant_property',
        value: 'some_id',
      },
      className: {
        type: 'constant_property',
        value: 'foo bar',
      },
    },
    children: [
      {
        id: 'child_1',
        type: 'html-element',
        name: 'span',
        children: [
          {
            id: 'child_1',
            type: 'html-element',
            name: 'span',
          },
        ],
      },
      {
        id: 'child_2',
        type: 'html-element',
        name: 'span',
        children: [
          {
            id: 'child_1',
            type: 'html-element',
            name: 'span',
          },
        ],
      },
    ],
  };
  const { container } = render(<HTMLNodeRender node={node} ctx={dummyCTX} />);

  expect(container).toMatchSnapshot();
});
