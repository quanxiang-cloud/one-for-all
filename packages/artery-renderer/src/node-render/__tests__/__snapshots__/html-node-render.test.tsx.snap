// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLNodeRender_match_snapshots 1`] = `
<div>
  <div
    class="foo bar"
    id="some_id"
  >
    <span>
      <span />
    </span>
    <span>
      <span />
    </span>
  </div>
</div>
`;

exports[`HTMLNodeRender_should_handle_empty_html_tag 1`] = `<div />`;
